# Homework 2 Programming Problem 8 (30 points)

## Problem Description

In this problem you will use linear least squares to fit a linear function to a 3D temperature field, with x,y,z locations and an associated temperature T.

Fill out the notebook as instructed, making the requested plots and printing necessary values. 

*You are welcome to use any of the code provided in the lecture activities.*

#### Summary of deliverables:

Results:
- Predicted temperature `T(5,5,5)` using a hand-coded LLS squares model with a linear function
- Direction of travel from `(5,5,5)` to experience the greatest decrease in temperature

Discussion:
- Reasoning for how we can use our fitted function to determine the direction of greatest decrease in temperature

#### Imports and Utility Functions:

In [5]:
import numpy as np
import matplotlib.pyplot as plt

## Load the data

The data is contained in `tempfield.npy` and can be loaded with `np.load(tempfield.npy)`. The first three columns correspond to the `x`, `y`, and `z` locations of the data points, and the 4th column corresponds to the temperature `T` at the respective point. Store the data as you see fit.

In [18]:
# YORU CODE GOES HERE
data = np.load("tempfield.npy")

x = data[:,0]
y = data[:,1]
z = data[:,2]
target = data[:,3]

## LLS Regression in 3D

Now fit a linear function to the data using the closed for of LLS regression. Use your fitted function to report the predicted temperature at `x = 5, y = 5, z = 5`. You are free to add regularization to your model, but this is not required and will not be graded.

In [19]:
X = np.concatenate([x.reshape(-1,1),y.reshape(-1,1),z.reshape(-1,1), np.ones_like(x).reshape(-1,1)], axis=1)
w = np.linalg.inv(X.T @ X) @ X.T @ target.reshape(-1,1)

In [20]:
# Temperature at x,y,z = 5,5,5 is 
print("Temperature at x,y,z = 5,5,5 is ", w[0]*5 + w[1]*5 + w[2]*5 + w[3])

Temperature at x,y,z = 5,5,5 is  [21.38907338]


## Gradient Intuition

Using the function you fit in the previous part, in which direction should one move from the point `p = (5,5,5)` to experience the largest decrease in temperature in the immediate neighborhood of the point? Report the specific direction, along with your reasoning.

In [36]:
# The gradient at x,y,z = 5,5,5 is 
# The final function can be tareget = w[0]x + w[1]y + w[2]z + w[3] * bias
# The gradiant is 
print("The inverse gradient at x,y,z = 5,5,5 is ", w[0], w[1], w[2])


The gradient at x,y,z = 5,5,5 is  [-0.13421269] [0.11741862] [0.72677374]


As above mention, the graient at (5,5,5) is [-0.13421269, 0.11741862, 0.72677374], so the direction of greatest decrease in temperature is [0.13421269, -0.11741862, -0.72677374]. The reason is that the gradient points in the direction of steepest ascent, and so the negative of the gradient points in the direction of steepest descent